### PR TITLE
tmedia 689 autoplay carousel

### PR DESCRIPTION
**Pull request by JackHowa**
## Ticket

- [TMEDIA-689](https://arcpublishing.atlassian.net/browse/TMEDIA-689) sub-task of https://arcpublishing.atlassian.net/jira/software/c/projects/TMEDIA/boards/875?modal=detail&selectedIssue=TMEDIA-689&assignee=5e387d6a1b1d910e5dfd7a9b

## Description

Add autoplay and custom button functionality

## Acceptance Criteria

- Autoplay functionality opt-in and -out on arrival to the page of the carousel
- Feature is accessible by detecting reduced motion 

## Test Steps

1. Checkout branch - `git checkout TMEDIA-689-autoplay-gallery`
2. Run Storybook `npm run storybook`
3. Look at the storybook for autoplay functionality 
Screen Shot 2022-04-18 at 17 00 03
4. Notice "Start Autoplay" on first arrival to a story with `enableAutoplay`
5. Click `Start Autoplay`
6. Wait 4 seconds (default from engine theme sdk). Then you will go to the next slide. You will also see the "Stop Autoplay" available 
Screen Shot 2022-04-18 at 17 01 32
8. Go to the custom button story example to see the buttons with our own custom styling work as before 
Screen Shot 2022-04-18 at 17 02 44

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed relevant documentation has been updated/added.
- [x] Add label - **ready for review** when the pull request is ready for someone to begin reviewing

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] All GitHub Actions pass
- [ ] Ran the code locally based on the test instructions.
- [ ] Checked Chromatic for Storybook changes, accepted the updates if acceptable
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.
- [ ] Approve and Add label - **ready to merge** if you are happy with the pull request
- [ ] Want another reviewer? Add the label **additional review**

-----

## Notes 

- Checking for reduced motion per accessibility requirement https://www.w3.org/TR/wai-aria-practices-1.2/examples/carousel/carousel-1-prev-next.html

## Questions 

- The engine theme sdk does not let you set the interval duration. I didn't implement this feature but set the 4000 from engine theme sdk interval along with its same logic. Is this ok to leave out? 
- Should we take in autoplay phrases as well as custom buttons? I think that the buttons should be enough customization. Thoughts? This follows the same pattern as the next and previous and fullscreen buttons. 

## Todos 

- [x] Take in autoplay components in storybook
- [x] Include autoplay custom buttons in proptypes 
- [x] Describe feature in carousel docs
- [x] Add tests. Include mocked reduced motion 
- [x] Test clicking to initiate the autoplay

### CODE REVIEW

File: src/components/carousel/index.jsx [hunk 1]
```diff
@@ -115,6 +142,12 @@ const Carousel = ({
 		setPosition(position - 100 / slidesToShow);
 	};
 
+	// a prefers-reduced-motion user setting must always override Autoplay
+	const AutoplayEnabledAndAllowed =
+		enableAutoplay && !!window.matchMedia("'(prefers-reduced-motion: reduce)");
```
Comment (matthewroach) [inline src/components/carousel/index.jsx]: `window` will throw an error on Fusion server side render and cause this code to become client side only. We should add an `isServerSide()` util function to handle checking for `window` being defined

File: src/components/carousel/index.jsx [hunk 2]
```diff
@@ -182,6 +217,28 @@ const Carousel = ({
...
+			`${COMPONENT_CLASS_NAME}__button--stop-auto-play`,
+			toggleAutoplay
+		)
+	) : (
+		<DefaultStopAutoplayButton id={id} onClick={toggleAutoplay} />
+	);
+
 	/* istanbul ignore next  */
 	const fullScreenEnabledAllowed =
 		(document.fullscreenEnabled || document.webkitFullscreenEnabled) && enableFullScreen;
```
Comment (matthewroach) [inline src/components/carousel/index.jsx]: `document` is not available on Fusion server side render and will throw an error as per the `window` reference above

**Review by matthewroach — DISMISSED** [review]
We also need to add `aria-live="off"` to the track `div` when auto play is running, and set it to `aria-live="polite"` when not in auto play mode

File: src/components/carousel/index.jsx (hunk 2 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx]: @matthewroach updated. thanks for the note

File: src/components/carousel/index.jsx (hunk 1 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx]: @matthewroach updated. thanks for the note

**Comment by JackHowa** [comment]
@matthewroach updated based on feedback

File: src/components/carousel/index.jsx [hunk 3]
```diff
@@ -203,6 +264,8 @@ const Carousel = ({
 						{pageCountPhrase(slide, totalSlides) || `${slide} of ${totalSlides}`}
 					</p>
 				) : null}
+				{AutoplayEnabledAndAllowed && !isAutoplaying ? resolvedStartAutoplayButton : null}
```
Comment (matthewroach) [inline src/components/carousel/index.jsx]: Using two buttons - one for each interaction will require us to manage state and focus for the user - could we get away with using one and have the user set the label and icon?

Comment (JackHowa) [reply src/components/carousel/index.jsx]: I was following the same custom button pattern as the next and previous buttons. What if the user wants to put an icon on the left or right side of the label? or do something else custom? should I add an icon left or right option as well? 

Ideally, we'd want a custom block developer to use our button component kind of like how I've done in the storybook

new props: 
- autoplayStartLabelText
- autoplayStartIcon
- autoplayStartIconLocation 
- autoplayStopLabelText
- autoplayStopIcon
- autoplayStopIconLocation

File: src/components/carousel/index.stories.mdx [hunk 4]
```diff
@@ -362,3 +364,66 @@ const Feature = () => (
...
+			</Carousel.Item>
+			<Carousel.Item label="Slide 3 of 4">
+				<Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
+			</Carousel.Item>
+			<Carousel.Item label="Slide 4 of 4">
+				<Image src="/coffee.jpeg" alt="A cup of coffee" />
+			</Carousel.Item>
+		</Carousel>
+	</Story>
+</Canvas>
```
Comment (brentmiller1973) [inline src/components/carousel/index.stories.mdx:571]: @JackHowa Its important, especially considering the complexity of this component to see the component with with multiple states enabled.  This will really help QA as well.  For autoplay, I want to see examples when their are multiple slides visible and when it's in fullscreen mode.

```
** Autoplay enabled with full screen**

<Canvas>
    <Story name="Autoplay enabled with full screen">
        <Carousel id="carousel-1" label="Carousel of Images" slidesToShow={1} enableAutoplay enableFullScreen>
            <Carousel.Item label="Slide 1 of 4">
                <a href="/">
                    <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                </a>
            </Carousel.Item>
            <Carousel.Item label="Slide 2 of 4">
                <Image src="/camera.jpeg" alt="A camera with photos in front of it" />
            </Carousel.Item>
            <Carousel.Item label="Slide 3 of 4">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
            <Carousel.Item label="Slide 4 of 4">
                <Image src="/coffee.jpeg" alt="A cup of coffee" />
            </Carousel.Item>
        </Carousel>
    </Story>
</Canvas>

** Autoplay enabled with multiple slides **

<Canvas>
    <Story name="Autoplay enabled with multiple slides">
        <Carousel
            id="carousel-1"
            label="Carousel of Images"
            enableAutoplay
            nextButton={
                <Carousel.Button id="carousel-1" label="Next Slide">
                    <Icon name="ArrowRight" />
                </Carousel.Button>
            }
            previousButton={
                <Carousel.Button id="carousel-1" label="Previous Slide">
                    <Icon name="ArrowLeft" />
                </Carousel.Button>
            }
        >
            <Carousel.Item label="Slide 1 of 6">
                {({ viewable }) => (
                    <a href="/" aria-hidden={viewable ? null : true} tabIndex={viewable ? null : "-1"}>
                        <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                    </a>
                )}
            </Carousel.Item>
            <Carousel.Item label="Slide 2 of 6">
                <Image src="/camera.jpeg" alt="A camera with photos in front of it" />
            </Carousel.Item>
            <Carousel.Item label="Slide 3 of 6">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
            <Carousel.Item label="Slide 4 of 6">
                <Image src="/coffee.jpeg" alt="A cup of coffee" />
            </Carousel.Item>
            <Carousel.Item label="Slide 5 of 6">
                {({ viewable }) => (
                    <div>
                        <a href="/" aria-hidden={viewable ? null : true} tabIndex={viewable ? null : "-1"}>
                            <Image src="/computer-user.jpeg" alt="A person sat down using a laptop computer" />
                        </a>
                    </div>
                )}
            </Carousel.Item>
            <Carousel.Item label="Slide 6 of 6">
                <Image src="/canyon.jpeg" alt="Landscape view of a canyon" />
            </Carousel.Item>
        </Carousel>
    </Story>
</Canvas>
```

File: src/components/carousel/index.jsx [hunk 5]
```diff
@@ -182,9 +218,34 @@ const Carousel = ({
...
+		resolvedButton(
+			stopAutoplayButton,
+			id,
+			`${COMPONENT_CLASS_NAME}__button--stop-auto-play`,
+			toggleAutoplay
+		)
+	) : (
+		<DefaultStopAutoplayButton id={id} onClick={toggleAutoplay} />
+	);
+
```
Comment (brentmiller1973) [inline src/components/carousel/index.jsx:300]: @JackHowa I don't know with this is scope creep (I think it may be), but I would think it would be prudent that after the last slide is reached, the button text stops staying "Stop Autoplay" and instead says "Start Autoplay".  If it's pressed it can either start back at slide one or go in reverse....up to product.

**Review by brentmiller1973 — DISMISSED** [review]
Looks great Jack!  Left two comments....the one on additional stories I think is a must, but the comment on what to do when the last slide is reached might be outside the scope of this ticket and may require some discussions with design and product.

File: src/components/carousel/index.jsx (hunk 3 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx]: I'm not totally sure how I could avoid using state here. what would determine whether to show the autoplayStartLabelText vs autoplayStopLabelText ? is this an instance of using aria attributes? @matthewroach 

looking to https://www.w3.org/TR/wai-aria-practices-1.2/examples/carousel/carousel-1-prev-next.html for inspiration

File: src/components/carousel/index.jsx (hunk 5 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx:300]: cool asked jenae in the ticket https://arcpublishing.atlassian.net/browse/TMEDIA-835?focusedCommentId=736246

File: src/components/carousel/index.stories.mdx (hunk 4 above)
Comment (JackHowa) [reply src/components/carousel/index.stories.mdx:571]: gotcha, can add to storybook

Comment (JackHowa) [reply src/components/carousel/index.stories.mdx:571]: updated @visionbegin

File: index.js [hunk 6]
```diff
@@ -34,12 +36,14 @@ export {
...
 	Input,
+	isServerSide,
 	Link,
 	MediaItem,
 	Overline,
 	Paragraph,
 	Pill,
 	Separator,
 	Stack,
+	useInterval,
```
Comment (nschubach) [inline index.js:47]: Should we namespace utility functions or leave them intermixed with the components?

Comment (JackHowa) [reply index.js:47]: in the past, we've intermixed them

Comment (JackHowa) [reply index.js:47]: I don't really have an issue with doing so. makes exports easier

File: src/components/carousel/index.jsx (hunk 3 above)
Comment (matthewroach) [reply src/components/carousel/index.jsx]: I don't think we need so many options - what if we provide the ability for them to set the label and icon for each state for now?

File: index.js (hunk 6 above)
Comment (nschubach) [reply index.js:47]: if we do, it would have to be another story, going to resolve this conversation.

File: src/components/carousel/index.jsx (hunk 3 above)
Comment (nschubach) [reply src/components/carousel/index.jsx]: Are you thinking of making the toggle button a callback where we pass in the current state and letting them decide?

File: src/components/carousel/index.jsx (hunk 5 above)
Comment (nschubach) [reply src/components/carousel/index.jsx:300]: Considering this works, could we discuss this and make it another story?  As far as I'm concerned, this is the only thing holding this up.

File: src/components/carousel/index.jsx (hunk 3 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx]: yep can do @matthewroach 

> I don't think we need so many options - what if we provide the ability for them to set the label and icon for each state for now?

**Comment by matthewroach** [comment]
@JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

**Comment by matthewroach** [comment]
@JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

**Comment by JackHowa** [comment]
@matthewroach will work on it; thanks for the note

**Comment by JackHowa** [comment]
> @JackHowa Looking through the pull requests for the updates to the carousel with the additional controls above - has there been a change in design or product decision to order the controls in this new way vs the current design of Full Screen and Autoplay on the left, and label and additional buttons on the right?

@matthewroach updated the order and their containers. 

Screen Shot 2022-04-20 at 09 23 11

-  [x] Will work on the feature change of only labels and icons for buttons rather than custom buttons and a fallback like next/previous
- [x] Oh I see. The autoplay does not immediately toggle off in the button when it hits the end. you have to wait until the next 4 seconds interval for that to happen. Will update autoplay stop functionality

File: src/components/carousel/index.jsx (hunk 5 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx:300]: @visionbegin this has been updated per the ticket. At the end of the cycle (after the 4 second interval like engine theme sdk), the autoplay will stop and stay at the last slide

Comment (JackHowa) [reply src/components/carousel/index.jsx:300]: https://github.com/WPMedia/arc-themes-components/pull/73/commits/4fced71e28ff79cc6f10df8d089dab372fe0df9f

**Comment by JackHowa** [comment]
@matthewroach 

like this? 

```jsx
const AutoplayButton = ({ id, onClick, iconNode, text, ariaLabel }) => (
	<Button
		id={id}
		onClick={onClick}
		label={ariaLabel}
		className={`${COMPONENT_CLASS_NAME}__button ${COMPONENT_CLASS_NAME}__button--auto-play`}
	>
		{iconNode}
		{text}
	</Button>
);
```

**Review by nschubach — APPROVED** [review]
👍

**Comment by JackHowa** [comment]
@matthewroach updated to align items within the controls container next to each other horizontally rather than vertically by default

File: src/components/carousel/index.jsx [hunk 7]
```diff
@@ -107,15 +126,17 @@ const Carousel = ({
 	showAdditionalSlideControls,
 	showLabel,
 	slidesToShow,
-	fullScreenShowButton,
-	fullScreenMinimizeButton,
-	enableFullScreen,
```
Comment (JackHowa) [inline src/components/carousel/index.jsx:112]: updating alphabetical

File: src/components/carousel/index.jsx (hunk 5 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx:300]: @visionbegin this has been updated. is it ok to resolve?

File: src/components/carousel/index.jsx (hunk 3 above)
Comment (JackHowa) [reply src/components/carousel/index.jsx]: this has been updated @matthewroach

**Review by ecarlisle — APPROVED** [review]
@JackHowa :  I think the code, testing, and Storybook examples are all great! The only question I had was when Start Autoplay was selected on the last slide. I see that's been brought up and you've asked Product about it here:
https://arcpublishing.atlassian.net/browse/TMEDIA-835?focusedCommentId=736246
Great job!

**Comment by JackHowa** [comment]
@visionbegin saw you had ❤️ the post in slack about reviewing the pr. but will merge this for now as far as two approvals. I reached out with the comments and resolving those. I think they've been addressed so I will resolve them
